### ISSUE TITLE
spacy.en.English() gets hanged

### ISSUE THREAD

**Issue by ciah0704** [honnibal/spaCy#218]
Hi! Congratulations on the awesome work! :) 
I am trying to run spacy using spacy.en.English(). But it gets stuck there. Does not move forward. I am using ubuntu 15. Please help me out.

**Comment by honnibal** [comment]
Hmm. Two thoughts:

1. How much memory do you have available? It needs up to 4gb
2. How long did you wait? It needs 1-2 minutes to load

**Comment by ciah0704** [comment]
Oh okay! memory is not an issue. It takes around 2-3 minutes. I guess its okay then. But still, any suggestions on making it run quicker?

**Comment by honnibal** [comment]
I'm working on a neural network model that should take much less memory.

You can do set the parser=False flag, to avoid loading the parser model. This will make it load much faster, but doesn't let you use a lot of the functionality.

**Comment by Snaipe** [comment]
@honnibal how much less memory are we talking? Currently the model takes around 1.1 GB of memory on my end, which makes it impossible to host on a small cloud worker, but if it can be reduced along with the load time, that would really be awesome.

Any ETA on when the neural network model hits a release?

**Comment by honnibal** [comment]
I've been doing nothing else for about 4 or 5 weeks now, which is why I've been a bit quiet :). I thought I'd get this done much quicker, but I took a detour to understand the latest papers.

I've got the NN implemented and it's passing basic tests, and doing what it should on binary text classification. I'm currently doing the POS tagging experiments. There seem to be some bugs in this part. I hope to have the tagger performing decently tonight, and maybe the parser doing useful things by the end of the week.

It's hard to say, though. Maybe there'll be some hold-ups. I'd be surprised if the NN model requires more than 500mb of memory. I think we might even be able to get it under 200mb.

**Comment by dandxy89** [comment]
What library are you you working with @honnibal or is it all custom?

**Comment by honnibal** [comment]
It's all custom. Current messy state here: https://github.com/honnibal/thinc/blob/tmp_nn2/thinc/nn.pxd

The batch normalization stuff isn't working yet, and there's a fair bit of code duplication while I've been playing around. I expect the version that gets linked to spaCy in the end will be about 500 lines. Everything's implemented as a staticmethod, because I liked namespacing them that way. They then become inlined to work around a limitation in Cython that I really should report.

So, why not Theano/Tensor Flow/Caffe etc? It's not just NIH, I promise :). There are a few things about our problem that make these libraries a poor fit.

The main difference is that the parser can't batch up problems and send them to a library. The parser is a push-down automaton, and the model has to determine which action to take next. The the parser has to get that action, and perform it. Then we can extract features from the state and advance. The loop looks like this:

    state = ParseState(sentence, tags)
    while not state.is_final():
        valid_actions = transition_system.check_valid(state)
        features = extract_features(state)
        scores = model.predict(features)
        action = arg_max_if_true(scores, is_valid)
        transition_system.apply(action, state)

The other consideration is that I really don't want to give up my hash tables. NLP problems are very sparse, so modelling everything as a dense matrix doesn't work so well.

So, I wanted to achieve the following:

1. The ML functions must not require the GIL. This way, we can finally properly multi-thread spaCy.
2. No system dependencies, e.g. depending on Eigen, Boost, Caffe etc is unacceptable. We're really serious about making spaCy's installation bullet-proof (next release is a massive advance on that). We can't achieve that if we rely on system libraries.
3. You shouldn't need to pre-define how many words are in your vocabulary. During training, it should just grow the hash table.
4. I want to be able to define group L1 penalties flexibly, to e.g. prefer models that don't fine-tune many word embeddings. This will allow us to re-use the same embeddings between tasks.
5. I want to be able to look over the code and understand it. We don't need so many abstractions, because we don't need to create some sort of "pluggable architecture". We can grow out the abstractions in time if we want. But for now the best way to ensure future changes can be integrated easily is to keep the code short and direct. If the library is <1,000 lines, integrating the next paper to be released can't be much work. But if we commit to a particular abstraction, it might be.

Note that I'm making a conscious trade-off here: at this point I'm not looking to support any other users of this code, other than spaCy, with this library. That buys a lot of freedom.

**Comment by dandxy89** [comment]
@honnibal Thanks for your response and appreciate the time that you put into responding to my question. 

I plan to look into your implementation of hash tables and the current nn file above you highlighted.
